src/i2c_software.c

```diff
@@ -0,0 +1,164 @@
+// Software I2C emulation
+//
+// Copyright (C) 2023  Kevin O'Connor <kevin@koconnor.net>
+// Copyright (C) 2023  Alan.Ma <tech@biqu3d.com>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <string.h> // memcpy
+#include "board/gpio.h" // gpio_out_setup
+#include "board/internal.h" // gpio_peripheral
+#include "board/misc.h" // timer_read_time
+#include "basecmd.h" // oid_alloc
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // sched_shutdown
+#include "i2ccmds.h" // i2cdev_set_software_bus
+
+struct i2c_software {
+    struct gpio_out scl, sda;
+    uint32_t sda_pin;
+    uint8_t addr;
+    unsigned int ticks;
+};
+
+void
+command_i2c_set_software_bus(uint32_t *args)
+{
+    struct i2cdev_s *i2c = i2cdev_oid_lookup(args[0]);
+    struct i2c_software *is = alloc_chunk(sizeof(*is));
+    // Max speed = 1MHz
+    unsigned int speed = args[3] > 1000000 ? 1000000 : args[3];
+    is->ticks = 1000000000 / speed / 2;
+    is->addr = (args[4] & 0x7f) << 1; // address format shifted
+    is->sda_pin = args[2];
+    is->scl = gpio_out_setup(args[1], 1);
+    is->sda = gpio_out_setup(is->sda_pin, 1);
+    i2cdev_set_software_bus(i2c, is);
+}
+DECL_COMMAND(command_i2c_set_software_bus,
+             "i2c_set_software_bus oid=%c scl_pin=%u sda_pin=%u"
+             " rate=%u address=%u");
+
+static unsigned int
+nsecs_to_ticks(uint32_t ns)
+{
+    return timer_from_us(ns * 1000) / 1000000;
+}
+
+void i2c_delay(unsigned int n) {
+    unsigned int t = timer_read_time() + nsecs_to_ticks(n);
+    while (t > timer_read_time());
+}
+
+static void
+i2c_software_send_ack(struct i2c_software *is, const uint8_t ack)
+{
+    gpio_out_write(is->sda, ack);
+    i2c_delay(is->ticks);
+    gpio_out_write(is->scl, 1);
+    i2c_delay(is->ticks);
+    gpio_out_write(is->scl, 0);
+}
+
+static uint8_t
+i2c_software_read_ack(struct i2c_software *is)
+{
+    uint8_t nack = 0;
+    struct gpio_in sda_in = gpio_in_setup(is->sda_pin, 1);
```

Comment:
It is not valid to call `gpio_in_setup()` nor `gpio_out_setup()` multiple times.  (It may work on some architectures, but will not work on others.)  Use `gpio_out_reset()` and `gpio_in_reset()` instead.  See `src/tmcuart.c` as an example.

Reply:
Use `gpio_**_reset` instead of `gpio_**_setup`, just like in `src/tmcuart.c` .
https://github.com/Klipper3d/klipper/pull/6141/commits/55b5e928f03b6cd9d590c40cd11cea4d2e88c026